lib/msprime.c

```diff
@@ -6044,75 +5934,73 @@ static int MSP_WARN_UNUSED
 msp_beta_common_ancestor_event(msp_t *self, population_id_t pop_id, label_id_t label)
 {
     int ret = 0;
-    uint32_t j, n, max_pot_size;
-    avl_tree_t *ancestors, Q[5]; /* MSVC won't let us use num_pots here */
-    avl_node_t *x_node, *y_node, *q_node;
-    segment_t *x, *y;
-    double beta_x;
+    uint32_t j, n, num_participants;
+    avl_tree_t *ancestors, Q[4]; /* MSVC won't let us use num_pots here */
+    double beta_x, u, increment;
 
-    for (j = 0; j < 5; j++){
+    for (j = 0; j < 4; j++){
         avl_init_tree(&Q[j], cmp_segment_queue, NULL);
     }
     ancestors = &self->populations[pop_id].ancestors[label];
-    /* Choose x and y */
     n = avl_count(ancestors);
-    j = (uint32_t) gsl_rng_uniform_int(self->rng, n);
-    x_node = avl_at(ancestors, j);
-    assert(x_node != NULL);
-    x = (segment_t *) x_node->item;
-    avl_unlink_node(ancestors, x_node);
-    j = (uint32_t) gsl_rng_uniform_int(self->rng, n - 1);
-    y_node = avl_at(ancestors, j);
-    assert(y_node != NULL);
-    y = (segment_t *) y_node->item;
-    avl_unlink_node(ancestors, y_node);
-    self->num_ca_events++;
-    msp_free_avl_node(self, x_node);
-    msp_free_avl_node(self, y_node);
-
-    q_node = msp_alloc_avl_node(self);
-    if (q_node == NULL) {
-        ret = MSP_ERR_NO_MEMORY;
-        goto out;
-    }
-    avl_init_node(q_node, x);
-    q_node = avl_insert_node(&Q[4], q_node);
-
-    q_node = msp_alloc_avl_node(self);
-    if (q_node == NULL) {
-        ret = MSP_ERR_NO_MEMORY;
-        goto out;
-    }
-    avl_init_node(q_node, y);
-    q_node = avl_insert_node(&Q[4], q_node);
-
     beta_x = ran_inc_beta(self->rng,
                  2.0 - self->model.params.beta_coalescent.alpha,
                  self->model.params.beta_coalescent.alpha,
-                 self->model.params.beta_coalescent.truncation_point,
-                 self->model.params.beta_coalescent.acceptance_rate);
-
-    ret = msp_multi_merger_common_ancestor_event(self, beta_x, ancestors, Q);
-    if (ret < 0) {
-        goto out;
+                 self->model.params.beta_coalescent.truncation_point);
+
+    /* We calculate the probability of accepting the event */
+    if (beta_x > 1e-9) {
+        u = (n - 1) * log(1 - beta_x)
```

Comment:
We'll need to make sure that the C tests are hitting both of these paths. Might take some parameter tweaking.

Reply:
I've changed one of the tests C test cases to have `alpha` near 1, and the other near 2. They will go down the two different paths with rare exceptions.

Reply:
Yep, both cases are covered. Great.